tests/UnitTests/Foundation/DispatchTests.mm

```diff
@@ -28,173 +28,180 @@
+
+    EXPECT_EQ(context2, dispatch_queue_get_specific(queue, key));
+    EXPECT_EQ(context, dispatch_queue_get_specific(queue, key2));
+
+    dispatch_queue_set_specific(queue, key3, context3, nullptr);
 
-	dispatch_queue_set_specific(queue, key3, context3, nullptr);
-    
     dispatch_group_async(group, queue, ^{
-		EXPECT_EQ(context2, dispatch_get_specific(key));
-		EXPECT_EQ(context, dispatch_get_specific(key2));
-		EXPECT_EQ(context3, dispatch_get_specific(key3));
+        EXPECT_EQ(context2, dispatch_get_specific(key));
+        EXPECT_EQ(context, dispatch_get_specific(key2));
+        EXPECT_EQ(context3, dispatch_get_specific(key3));
     });
-	ASSERT_EQ(0, dispatch_group_wait(group, _createEndTestTime()));
+    ASSERT_EQ(0, dispatch_group_wait(group, _createEndTestTime()));
 
-	dispatch_release(queue);
-	dispatch_release(group);
+    dispatch_release(queue);
+    dispatch_release(group);
 }
 
 TEST(DispatchTests, SpecificDestructor) {
-	dispatch_queue_t queue = dispatch_queue_create("queue", nullptr);
-	dispatch_group_t group = dispatch_group_create();
+    dispatch_queue_t queue = dispatch_queue_create("queue", nullptr);
+    dispatch_group_t group = dispatch_group_create();
+    destructorSemaphore = dispatch_semaphore_create(0);
 
     dispatch_queue_set_specific(queue, key, context, destructor);
 
-	EXPECT_EQ(0, destructorCount);
-    
+    EXPECT_EQ(0, destructorCount);
+
     dispatch_group_async(group, queue, ^{
-		EXPECT_EQ(dispatch_get_specific(key), context);
+        EXPECT_EQ(dispatch_get_specific(key), context);
         dispatch_queue_set_specific(queue, key, context2, destructor);
-		EXPECT_EQ(1, destructorCount);
-    });
-	ASSERT_EQ(dispatch_group_wait(group, _createEndTestTime()), 0);
-    
-	EXPECT_EQ(context2, dispatch_queue_get_specific(queue, key));
-	dispatch_queue_set_specific(queue, key, context, destructor);
-	EXPECT_EQ(2, destructorCount);
-    
+        // Destructor can be called asynchrounously. Wait for deestructor to be called.
+        dispatch_semaphore_wait(destructorSemaphore, _createEndTestTime());
+        EXPECT_EQ(1, destructorCount);
+    });
+    ASSERT_EQ(dispatch_group_wait(group, _createEndTestTime()), 0);
+
+    EXPECT_EQ(context2, dispatch_queue_get_specific(queue, key));
+    dispatch_queue_set_specific(queue, key, context, destructor);
+    dispatch_semaphore_wait(destructorSemaphore, _createEndTestTime());
+    EXPECT_EQ(2, destructorCount);
+
     dispatch_group_async(group, queue, ^{
-		EXPECT_EQ(context, dispatch_get_specific(key));
-		EXPECT_EQ(2, destructorCount);
+        EXPECT_EQ(context, dispatch_get_specific(key));
+        EXPECT_EQ(2, destructorCount);
     });
-	ASSERT_EQ(0, dispatch_group_wait(group, _createEndTestTime()));
+    ASSERT_EQ(0, dispatch_group_wait(group, _createEndTestTime()));
 
-	dispatch_release(queue);
-	dispatch_release(group);
+    dispatch_release(queue);
+    dispatch_release(group);
+    dispatch_release(destructorSemaphore);
```

Comment:
Please make sure to set this to null. Otherwise, subsequent tests will signal on a free’d semaphore!

I guess also check if the semaphore exists in the dtor callback ;p